hypothesis-python/tests/conjecture/test_ir_tree.py

```diff
@@ -0,0 +1,224 @@
+
+    # the copied leaf should have a value of new_value.
+    assert ir_value_eq(ir_type, new_value, leaf2.value)
+    # if we chose a value different from the original leaf, they should no
+    # longer be equal.
+    if not ir_value_eq(ir_type, leaf.value, new_value):
+        assert leaf != leaf2
+
+
+@given(leaves(was_forced=True))
+def test_copy_forced_leaf_with_value(leaf):
+    # we disallow modifying the value of forced leaves while copying for now.
+    new_value = draw_ir_value(leaf.ir_type, leaf.kwargs)
+    with pytest.raises(AssertionError):
+        leaf.copy(with_value=new_value)
+
+
+def test_copy_empty_ir_tree():
+    ir_tree = IRTree()
+    assert ir_tree.root is None
+
+    copied = ir_tree.copy()
+    assert ir_tree == copied
+    assert copied.root is None
+
+
+# avoid slowdowns or overruns
+@settings(max_examples=20, stateful_step_count=10)
+class CopyIRTreeIsEqual(RuleBasedStateMachine):
+    def __init__(self):
+        super().__init__()
+        self.data = fresh_data()
+
+    @rule(kwargs=draw_integer_kwargs())
+    def draw_integer(self, kwargs):
+        self.data.draw_integer(**kwargs)
+
+    @rule(kwargs=draw_string_kwargs())
+    def draw_string(self, kwargs):
+        self.data.draw_string(**kwargs)
+
+    @rule(kwargs=draw_boolean_kwargs())
+    def draw_boolean(self, kwargs):
+        self.data.draw_boolean(**kwargs)
+
+    @rule(kwargs=draw_bytes_kwargs())
+    def draw_bytes(self, kwargs):
+        self.data.draw_bytes(**kwargs)
+
+    @rule(kwargs=draw_float_kwargs())
+    def draw_float(self, kwargs):
+        self.data.draw_float(**kwargs)
+
+    @invariant()
+    def copying_ir_tree_is_equal(self):
+        assert self.data.ir_tree == self.data.ir_tree.copy()
+
+    @rule(data_strategy=st.data())
+    def invariant_copying_ir_tree_while_replacing_nodes(self, data_strategy):
+        # this function is morally an invariant, but is implemented as an @rule
+        # in order to access st.data() to draw a random leaf.
+        if not self.data.ir_tree.leaves():
+            return
+
+        tree = self.data.ir_tree
+        tree.update_locations()
+
+        leaves = tree.leaves()
+        i = data_strategy.draw(st.integers(0, len(leaves) - 1))
+        leaf = leaves[i]
+        new_value = draw_ir_value(leaf.ir_type, leaf.kwargs)
+
+        tree2 = tree.copy(
+            replacing_nodes=[(leaf.location, leaf.copy(with_value=new_value))]
+        )
+        leaves2 = tree2.leaves()
+
+        assert len(leaves) == len(leaves2)
+        assert ir_value_eq(leaf.ir_type, leaves2[i].value, new_value)
+
+        if not ir_value_eq(leaf.ir_type, leaf.value, new_value):
+            assert tree != tree2
+
+
+TestCopyIRTreeIsEqual = CopyIRTreeIsEqual.TestCase
+
+
+@given(leaves())
+def test_new_leaves_have_no_location(leaf):
+    assert leaf.location is None
+    assert leaf.depth is None
+    assert leaf.index_in_parent is None
+
+
+def test_leaves():
+    data = fresh_data()
```

Comment:
This is outside `@given()`, but without a seed it risks being flaky.

(I think in practice you're relying on our self-tests having seeded the PRNG somewhere, which we should also stop doing... but that's another issue)